src/lightgbm/src/main/scala/TrainUtils.scala

```diff
@@ -36,95 +36,172 @@ private object TrainUtils extends Serializable {
+      }
+
+    // Validate generated dataset has the correct number of rows and cols
+    validateDataset(datasetPtr.get)
+    addField(labels, "label", numRows, datasetPtr)
+    if (weightColumn.isDefined) {
+      val weights = rows.map(row => row.getDouble(row.fieldIndex(weightColumn.get)))
+      addField(weights, "weight", numRows, datasetPtr)
+    }
+    datasetPtr
+  }
+
+  def createBooster(trainParams: TrainParams, trainDatasetPtr: Option[SWIGTYPE_p_void],
+                    validDatasetPtr: Option[SWIGTYPE_p_void]): Option[SWIGTYPE_p_void] = {
+    // Create the booster
+    val boosterOutPtr = lightgbmlib.voidpp_handle()
+    val parameters = trainParams.toString()
+    LightGBMUtils.validate(lightgbmlib.LGBM_BoosterCreate(trainDatasetPtr.get, parameters, boosterOutPtr), "Booster")
+    val boosterPtr = Some(lightgbmlib.voidpp_value(boosterOutPtr))
+    if (trainParams.modelString != null && !trainParams.modelString.isEmpty) {
+      val booster = LightGBMUtils.getBoosterPtrFromModelString(trainParams.modelString)
+      LightGBMUtils.validate(lightgbmlib.LGBM_BoosterMerge(boosterPtr.get, booster), "Booster Merge")
+    }
+    if (validDatasetPtr.isDefined) {
+      LightGBMUtils.validate(lightgbmlib.LGBM_BoosterAddValidData(boosterPtr.get,
+        validDatasetPtr.get), "Add Validation Dataset")
+    }
+    boosterPtr
+  }
 
-      // Validate generated dataset has the correct number of rows and cols
-      validateDataset(datasetPtr.get)
-      addField(labels, "label", numRows, datasetPtr)
+  def saveBoosterToString(boosterPtr: Option[SWIGTYPE_p_void], log: Logger): String = {
+    val bufferLength = LightGBMConstants.defaultBufferLength
+    val bufferLengthPtr = lightgbmlib.new_longp()
+    lightgbmlib.longp_assign(bufferLengthPtr, bufferLength)
+    val bufferLengthPtrInt64 = lightgbmlib.long_to_int64_t_ptr(bufferLengthPtr)
+    val bufferOutLengthPtr = lightgbmlib.new_int64_tp()
+    lightgbmlib.LGBM_BoosterSaveModelToStringSWIG(boosterPtr.get, 0, -1, bufferLengthPtrInt64, bufferOutLengthPtr)
+  }
 
-      if (weightColumn.isDefined) {
-        val weights = rows.map(row => row.getDouble(row.fieldIndex(weightColumn.get)))
-        addField(weights, "weight", numRows, datasetPtr)
-      }
+  case class BestEvaluation(evalCounts: Int, evalNames: Array[String], bestScore: Array[Double],
+                            bestScores: Array[Array[Double]], bestIter: Array[Int])
 
-      var boosterPtr: Option[SWIGTYPE_p_void] = None
+  def createBestEval(boosterPtr: Option[SWIGTYPE_p_void]): Option[BestEvaluation] = {
+    // Need to keep track of best scores for each metric, see callback.py in lightgbm for reference
+    val evalCountsPtr = lightgbmlib.new_intp()
+    val resultCounts = lightgbmlib.LGBM_BoosterGetEvalCounts(boosterPtr.get, evalCountsPtr)
+    LightGBMUtils.validate(resultCounts, "Booster Get Eval Counts")
+    val evalCounts = lightgbmlib.intp_value(evalCountsPtr)
+    val bestScore = new Array[Double](evalCounts)
```

Comment:
use maps instead of mutation

Reply:
why are you retuning an empty array of arrays? why not make an array later when you need it

Reply:
"why are you retuning an empty array of arrays? why not make an array later when you need it"
done
"use maps instead of mutation"
for different indexes you can have different best score/iteration, and we reuse the same array throughout so it doesn't make sense to allocate it more than once